docroot/modules/custom/mass_content/src/EventsRendererOrgPages.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+namespace Drupal\mass_content;
+
+use Drupal\mayflower\Helper;
+use Drupal\paragraphs\ParagraphInterface;
+
+/**
+ * Render events for org_events paragraph.
+ */
+class EventsRendererOrgPages {
+
+  /**
+   * The Drupal\mass_content\EventManager service.
+   *
+   * @var \Drupal\mass_content\EventManager
+   */
+  private $eventManager;
+
+  /**
+   * Stores if there are past events.
+   *
+   * @var bool
+   */
+  private $hasPastEvents;
+
+  /**
+   * The parent node where the paragraph is referenced.
+   *
+   * @var \Drupal\node\Entity\Node
+   */
+  private $parentNode;
+
+  /**
+   * Creates an events renderer object.
+   */
+  public function __construct(EventManager $event_manager, ParagraphInterface $paragraph) {
+    $this->eventManager = $event_manager;
+    $this->paragraph = $paragraph;
+    $this->parentNode = Helper::getParentNode($paragraph);
+    $this->hasPastEvents = $this->eventManager->hasPast($this->parentNode);
+  }
+
+  /**
+   * Returns a render array with the upcoming or past events.
+   */
+  public function render() {
+    // Gather events associated with this node.
+    $eventManager = \Drupal::service('mass_content.event_manager');
+    $upcoming = $eventManager->hasUpcoming($this->parentNode);
+    $render = $upcoming ? $this->nextEvents() : $this->pastEventsLink();
+    // To update if any event changes.
+    $render['#cache']['tags'][] = 'node_list:event';
```

Comment:
🤔 Do we want to add the tag manually, or should we use the an existing API like https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Render%21Renderer.php/function/Renderer%3A%3AaddCacheableDependency/9.4.x or `CacheableMetadata`?

Reply:
I was more referring to something like:

```php
    (new CacheableMetadata())->addCacheTags(['node_list:event]')
      ->applyTo($render);
```

On the one hand, it's more verbose, but it also means it reduces strings you can typo from three (`#cache`, `tags`, `node_list:event`) to just `node_list:event`.

Let's keep it as is for this PR. Thanks for thinking through it!